### PR TITLE
Only rewrite valtree-constants to patterns and keep other constants opaque

### PR DESCRIPTION
**Pull request by oli-obk**
Now that we can reliably fall back to comparing constants with `PartialEq::eq` to the match scrutinee, we can

1. eagerly try to convert constants to valtrees
2. then deeply convert the valtree to a pattern
3. if the to-valtree conversion failed, create an "opaque constant" pattern.

This PR specifically avoids any behavioral changes or major cleanups. What we can now do as follow ups is

* move the two remaining call sites to `destructure_mir_constant` off that query
* make valtree to pattern conversion infallible
    * this needs to be done after careful analysis of the effects. There may be user visible changes from that.

based on https://github.com/rust-lang/rust/pull/111768

### CODE REVIEW

File: compiler/rustc_mir_build/src/thir/pattern/deconstruct_pat.rs [hunk 1]
```diff
@@ -141,22 +141,27 @@ impl IntRange {
...
+                        // could remove this branch and just fall through, which
+                        // is more general but much slower.)
+                        return scalar.to_bits_or_ptr_internal(target_size).unwrap().left();
+                    }
+                    mir::ConstantKind::Ty(c) => match c.kind() {
+                        ty::ConstKind::Value(valtree) => {
+                            return valtree.unwrap_leaf().to_bits(target_size).ok();
+                        }
+                        _ => {}
+                    },
```
Comment (Nadrieril) [inline compiler/rustc_mir_build/src/thir/pattern/deconstruct_pat.rs:158]: You could use an if-let guard here to avoid the function-wrapping

**Comment by BoxyUwU** [comment]
would also like to review this

File: compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs [hunk 2]
```diff
@@ -141,36 +172,78 @@ impl<'tcx> ConstToPat<'tcx> {
...
+                            lint::builtin::POINTER_STRUCTURAL_MATCH,
+                            self.id,
+                            self.span,
+                            PointerPattern,
+                        );
+                    }
+                    _ => {}
                 }
             }
         }
```
Comment (oli-obk) [inline compiler/rustc_mir_build/src/thir/pattern/const_to_pat.rs:241]: This whole nested conditions and lint emissions thing is fishy, but it reproduces the linting behaviour on master. I want to clean it up, but we first should decide on what linting behaviour we want at all, and that should happen in a separate PR

File: compiler/rustc_middle/src/mir/mod.rs [hunk 3]
```diff
@@ -2487,7 +2487,12 @@ impl<'tcx> ConstantKind<'tcx> {
         if let Some(lit_input) = lit_input {
             // If an error occurred, ignore that it's a literal and leave reporting the error up to
             // mir.
-            match tcx.at(expr.span).lit_to_mir_constant(lit_input) {
+            match tcx
+                .at(expr.span)
+                .lit_to_const(lit_input)
+                .map(Self::Ty)
+                .or_else(|_| tcx.at(expr.span).lit_to_mir_constant(lit_input))
```
Comment (oli-obk) [inline compiler/rustc_middle/src/mir/mod.rs]: remove this eagerly

File: compiler/rustc_mir_build/src/thir/pattern/mod.rs [hunk 4]
```diff
@@ -518,16 +518,22 @@ impl<'a, 'tcx> PatCtxt<'a, 'tcx> {
...
+        let cid = GlobalId { instance, promoted: None };
+        // Prefer
+        let const_value = self
+            .tcx
+            .const_eval_global_id_for_typeck(param_env_reveal_all, cid, Some(span))
+            .and_then(|val| match val {
+                Some(valtree) => Ok(mir::ConstantKind::Ty(self.tcx.mk_const(valtree, ty))),
+                None => self
+                    .tcx
+                    .const_eval_global_id(param_env_reveal_all, cid, Some(span))
```
Comment (oli-obk) [inline compiler/rustc_mir_build/src/thir/pattern/mod.rs:530]: unwrap here

File: compiler/rustc_mir_build/src/thir/pattern/mod.rs [hunk 5]
```diff
@@ -578,9 +584,21 @@ impl<'a, 'tcx> PatCtxt<'a, 'tcx> {
         span: Span,
     ) -> PatKind<'tcx> {
         let value = mir::ConstantKind::from_inline_const(self.tcx, anon_const.def_id);
```
Comment (oli-obk) [inline compiler/rustc_mir_build/src/thir/pattern/mod.rs:586]: make this return a ty::Const and simplify the logic below

**Review by lcnr — COMMENTED** [review]
r=me after nits

File: compiler/rustc_const_eval/src/const_eval/mod.rs [hunk 6]
```diff
@@ -131,38 +129,3 @@ pub(crate) fn try_destructure_mir_constant<'tcx>(
 
     Ok(mir::DestructuredConstant { variant, fields })
 }
-
-#[instrument(skip(tcx), level = "debug")]
-pub(crate) fn deref_mir_constant<'tcx>(
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/const_eval/mod.rs:136]: So happy to see this go. :)

Now, when will destructure_mir_constant disappear? ;)
It seems to be used only for pretty-printing, and for some obscure SIMD-related thing.

Comment (oli-obk) [reply compiler/rustc_const_eval/src/const_eval/mod.rs:136]: yea, the SIMD thing I wanna solve first, and then just move `destructure_mir_constant` into pretty printing and kill the query

**Pull request merged**